src/hotspot/os/posix/malloctrace/mallocTrace2.cpp

```diff
@@ -0,0 +1,2675 @@
+  if (nr_of_frames < 2) { \
+    frames[0] = real_func; \
+    frames[1] = (address) caller_address; \
+    nr_of_frames = MAX2(2, _max_frames); \
+  } \
+  if (_detailed_stats) { \
+    Atomic::add(&_stack_walk_time, Ticks::now().nanoseconds() - ticks); \
+    Atomic::add(&_stack_walk_count, (uint64_t) 1); \
+  }
+
+#if defined(ASSERT)
+static uint64_t ptr_hash_backup(void* ptr) {
+  uint64_t hash = (uint64_t) ptr;
+  hash = (~hash) + (hash << 21);
+  hash = hash ^ (hash >> 24);
+  hash = hash * 265;
+  hash = hash ^ (hash >> 14);
+  hash = hash * 21;
+  hash = hash ^ (hash >> 28);
+  hash = hash + (hash << 31);
+
+  return hash;
+}
+#endif
+
+static void after_child_fork() {
+  if (register_hooks != NULL) {
+    register_hooks(NULL);
+  }
+}
+
+
+bool MallocStatisticImpl::setup_hooks(registered_hooks_t* hooks, outputStream* st) {
+  if (register_hooks == NULL) {
+    register_hooks = (register_hooks_t*) dlsym((void*) RTLD_DEFAULT, REGISTER_HOOKS_NAME);
+    get_real_funcs = (get_real_funcs_t*) dlsym((void*) RTLD_DEFAULT, GET_REAL_FUNCS_NAME);
+
+    if ((register_hooks == NULL) || (get_real_funcs == NULL)) {
+      if (UseMallocHooks) {
+        st->print_raw_cr("Could not find preloaded libmallochooks while -XX:+UseMallocHooks is set. " \
+                         "This usually happens if the VM is not loaded via the JDK launcher (e.g. " \
+                         "java.exe). In this case you must preload the library by setting the " \
+                         "following environment variable: ");
+        print_needed_preload_env(st);
+      } else {
+        st->print_cr("Could not find preloaded libmallochooks. Try using -XX:+UseMallocHooks " \
+                     "Vm option to automatically preload it using the JDK launcher. Or you can set " \
+                     "the following environment variable: ");
+        print_needed_preload_env(st);
+      }
+
+      st->print_raw_cr("VM arguments:");
+      Arguments::print_summary_on(st);
+
+      return false;
+    }
+  }
+
+  _funcs = get_real_funcs();
+  register_hooks(hooks);
+
+  return true;
+}
+
+uint64_t MallocStatisticImpl::ptr_hash(void* ptr) {
+  if (!_track_free && (_to_track_mask == 0)) {
+    return 0;
+  }
+
+  uint64_t hash = (uint64_t) ptr;
+  hash = (~hash) + (hash << 21);
+  hash = hash ^ (hash >> 24);
+  hash = (hash + (hash << 3)) + (hash << 8);
+  hash = hash ^ (hash >> 14);
+  hash = (hash + (hash << 2)) + (hash << 4);
+  hash = hash ^ (hash >> 28);
+  hash = hash + (hash << 31);
+
+  assert(hash == ptr_hash_backup(ptr), "Must be the same");
+
+  return hash;
+}
+
+bool MallocStatisticImpl::should_track(uint64_t hash) {
+  if (_detailed_stats) {
+    if ((hash & _to_track_mask) < _to_track_limit) {
+      Atomic::add(&_tracked_ptrs, (uint64_t) 1);
+    } else {
+      Atomic::add(&_not_tracked_ptrs, (uint64_t) 1);
+    }
+  }
+
+  return (hash & _to_track_mask) < _to_track_limit;
+}
+
+#define malloc_not_suspended() (pthread_getspecific(_malloc_suspended) == NULL)
```

Comment:
The impact is small, since it just checks a thread-relative field for the pthread_getspecific() call.

Reply:
Ok, I've added a static bool field as a first check, since it move all the code which changes the key into a lock protected region, so we  don't have problem with concurrent modifications of the field.